conngen/conngen.cpp

```diff
@@ -37,185 +34,312 @@
+ * index-translated id of the first local neuron as the first
+ * entry. If this renders the range empty (i.e. because the first
+ * local id is beyond the last element of the range), the range is
+ * not added to the mask.
+ *
+ * \param masks The std::vector of Masks to populate
+ * \param sources The source ranges to create the source masks from
+ * \param targets The target ranges to create the target masks from
+ *
+ * \note Each process computes the full set of source and target
+ * masks, i.e. one mask per rank will be created on each rank.
+ *
+ * \note Setting the masks for all processes on each process might
+ * become a memory bottleneck when going to very large numbers of
+ * processes. Especially so for the source masks, which are all the
+ * same. This could be solved by making the ConnectionGenerator
+ * interface MPI aware and communicating the masks during connection
+ * setup.
+*/
 void
-nest::cg_connect( nest::ConnectionGeneratorDatum& cg,
-  IntVectorDatum& sources,
-  IntVectorDatum& targets,
-  const DictionaryDatum& params_map,
-  const Name& synmodel_name )
+cg_create_masks( std::vector< ConnectionGenerator::Mask >* masks,
+  RangeSet& sources,
+  RangeSet& targets )
 {
-  const Token synmodel =
-    kernel().model_manager.get_synapsedict()->lookup( synmodel_name );
-  if ( synmodel.empty() )
+  // The index of the left border of the currently looked at range
+  // (counting from 0). This is used for index translation.
+  size_t cg_idx_left = 0;
+
+  // For sources, we only need to translate from NEST to CG indices.
+  for ( RangeSet::iterator source = sources.begin(); source != sources.end();
+        ++source )
   {
-    throw UnknownSynapseType( synmodel_name.toString() );
+    size_t num_elements = source->last - source->first + 1;
+    size_t right = cg_idx_left + num_elements - 1;
+    for ( size_t proc = 0; proc
+            < static_cast< size_t >( kernel().mpi_manager.get_num_processes() );
+          ++proc )
+    {
+      ( *masks )[ proc ].sources.insert( cg_idx_left, right );
+    }
+    cg_idx_left += num_elements;
   }
-  const index synmodel_id = static_cast< index >( synmodel );
 
-  RangeSet source_ranges;
-  cg_get_ranges( source_ranges, ( *sources ) );
+  // Reset the index of the left border of the range for index
+  // translation for the targets.
+  cg_idx_left = 0;
 
-  RangeSet target_ranges;
-  cg_get_ranges( target_ranges, ( *targets ) );
+  for ( RangeSet::iterator target = targets.begin(); target != targets.end();
+        ++target )
+  {
+    size_t num_elements = target->last - target->first + 1;
+    for ( size_t proc = 0; proc
+            < static_cast< size_t >( kernel().mpi_manager.get_num_processes() );
+          ++proc )
+    {
+      // Make sure that the range is only added on as many ranks as
+      // there are elements in the range, or exactly on every rank,
+      // if there are more elements in the range.
+      if ( proc <= num_elements - 1 )
```

Comment:
Why `p <= n-1` instead of the more common `p < n`?

Reply:
Fixed in 43e8d6abf.